File: src/delta_plot.c
/* delta_plot.c       2011-06-23 */

/* Copyright 2010-2011 Emmanuel Paradis */

/* This file is part of the R-package `ape'. */
/* See the file ../COPYING for licensing issues. */

#include <R.h>

void delta_plot(double *D, int *size, int *nbins, int *counts, double *deltabar)
{
	int x, y, u, v; /* same notation than in Holland et al. 2002 */
	int n = *size, nb = *nbins;
	double dxy, dxu, dxv, dyu, dyv, duv, A, B, C, delta;
	int xy, xu, xv, yu, yv, uv, i;

	for (x = 0; x < n - 3; x++) {
		xy = x*n - x*(x + 1)/2; /* do NOT factorize */
		for (y = x + 1; y < n - 2; y++, xy++) {
			yu = y*n - y*(y + 1)/2; /* do NOT factorize */
			dxy = D[xy];
			xu = xy + 1;
			for (u = y + 1; u < n - 1; u++, xu++, yu++) {
				uv = u*n - u*(u + 1)/2; /* do NOT factorize */
				dxu = D[xu]; dyu = D[yu];
				xv = xu + 1; yv = yu + 1;
				for (v = u + 1; v < n; v++, xv++, yv++, uv++) {
					dxv = D[xv]; dyv = D[yv]; duv = D[uv];
					A = dxv + dyu; B = dxu + dyv; C = dxy + duv;
					if (A == B && B == C) delta = 0; else while (1) {
						if (C <= B && B <= A) {delta = (A - B)/(A - C); break;}
						if (B <= C && C <= A) {delta = (A - C)/(A - B); break;}
						if (A <= C && C <= B) {delta = (B - C)/(B - A); break;}
						if (C <= A && A <= B) {delta = (B - A)/(B - C); break;}
						if (A <= B && B <= C) {delta = (C - B)/(C - A); break;}
						if (B <= A && A <= C) {delta = (C - A)/(C - B); break;}
					}
					/* if (delta == 1) i = nb - 1; else */
					i = delta * nb;
					counts[i] += 1;
					deltabar[x] += delta;
					deltabar[y] += delta;
					deltabar[u] += delta;
					deltabar[v] += delta;
				}
			}
		}
	}
}

File: R/delta.plot.R
## delta.plot.R (2010-01-12)

##   Delta Plots

## Copyright 2010 Emmanuel Paradis

## This file is part of the R-package `ape'.
## See the file ../COPYING for licensing issues.

delta.plot <- function(X, k = 20, plot = TRUE, which = 1:2)
{
    if (is.matrix(X)) X <- as.dist(X)
    n <- attr(X, "Size")
    if (n < 4) stop("need at least 4 observations")

    ## add a category for the cases delta = 1
    ans <- .C(delta_plot, as.double(X), as.integer(n),
              as.integer(k), integer(k + 1), double(n),
              NAOK = TRUE)
    counts <- ans[[4]]
    ## add the counts of delta=1 to the last category:
    counts[k] <- counts[k] + counts[k + 1]
    counts <- counts[-(k + 1)]

    delta.bar <- ans[[5]]/choose(n - 1, 3)

    if (plot) {
        if (length(which) == 2) layout(matrix(1:2, 1, 2))
        if (1 %in% which) {
            barplot(counts, space = 0, xlab = expression(delta[q]))
            a <- axTicks(1)
            axis(1, at = a, labels = a/k)
        }
        if (2 %in% which)
            plot(delta.bar, type = "h", ylab = expression(bar(delta)))
    }

    invisible(list(counts = counts, delta.bar = delta.bar))
}
